src/src/com/microsoft/aad/adal/AuthenticationContext.java

```diff
@@ -1603,112 +1646,98 @@ public RefreshItem(String keyInCache, AuthenticationRequest request, TokenCacheI
-                // acquireTokenSilent expects userid field from UserInfo
-                userId = request.getLoginHint();
-            }
-            String keyUsed;
-            TokenCacheItem item;
-            if (!useMultiResourceRefreshToken) {
-                // target refreshToken for this resource only. CacheKey will
-                // include the resourceId in the cachekey
-                Logger.v(TAG, "Looking for regular refresh token");
-                keyUsed = CacheKey.createCacheKey(request, userId);
-                item = mTokenCacheStore.getItem(keyUsed);
-            } else {
-                // check multiResource item in cache. Cache key
-                // will not include resourceId in the cache key string.
-                Logger.v(TAG, "Looking for Multi Resource Refresh token");
-                keyUsed = CacheKey.createMultiResourceRefreshTokenKey(request, userId);
-                item = mTokenCacheStore.getItem(keyUsed);
-            }
-
-            // Temporarily disable the family client id feature. 
-            final boolean isFamilyClientIdFeatureEnabled = false;
-            if (isFamilyClientIdFeatureEnabled) {
-                // If still cannot find an item, try with family clientid
-                // Disable this feature for ADFS authority
-                if (item == null || StringExtensions.IsNullOrBlank(item.getRefreshToken())) {
-                    if (!UrlExtensions.isADFSAuthority(StringExtensions.getUrl(mAuthority))) {
-                        Logger.v(TAG + methodName, "No refresh token found, trying to find family client id item.");
-                        item = findFamilyItemForUser(request.getUserIdentifierType(), userId);
-                    } else {
-                        Logger.v(TAG + methodName, "No refresh token found, skip the family client id item lookup for ADFS authority.");
+            Logger.v(TAG + methodName, "Look for resource specific refresh token in the cache.");
+            refreshItem = findResourceSpecificRefreshToken(request);
+            
+            // If regular RT is found, but it's not MRRT, shouldn't bother trying with FRT. 
+            // All the token issued by EvoSts is MRRT. If regular RT is not MRRT, it basically means
+            // it's not issued by EvoSts. FoCI feature is only visible to RT issued by evoSTS. 
+            if (refreshItem == null || refreshItem.mMultiResource == true) {
+                Logger.v(TAG + methodName, "Resource specific RT does not exist or it is also a MRRT, "
+                        + "look for MRRT.");
+                refreshItem = findMultiResourceRefreshToken(request);
+                
+                if (refreshItem == null || refreshItem.mFamilyToken == true) {
+                    Logger.v(TAG + methodName, "MRRT does not exist or MRRT is also a FRT, "
+                            + "look for FRT.");
+                    refreshItem = findFamilyRefreshToken(request);
```

Comment:
It is worthwhile to explain in comments as to why if we already found an FRT that we go look for  FRT again. Or may be it does not need to be looked up again.